src/mds/CDir.cc

```diff
@@ -2160,10 +2161,171 @@ class C_IO_Dir_Committed : public CDirIOContext {
+  mempool::mds_co::compact_set<mempool::mds_co::string> stale_items;
 };
 
+// This is not locked by mds_lock
+void CDir::_omap_commit_ops(int r, int op_prio, version_t version, bool _new, bufferlist &dfts,
+                            vector<dentry_key_t>& to_remove, vector<dentry_commit_item> &to_set,
+			    mempool::mds_co::compact_set<mempool::mds_co::string> &stales)
+{
+  dout(10) << __func__ << dendl;
+
+  if (r < 0) {
+    mdcache->mds->handle_write_error_with_lock(r);
+    return;
+  }
+
+  C_GatherBuilder gather(g_ceph_context,
+                         new C_OnFinisher(new C_IO_Dir_Committed(this, version),
+			 mdcache->mds->finisher));
+
+  SnapContext snapc;
+  object_t oid = get_ondisk_object();
+  object_locator_t oloc(mdcache->mds->mdsmap->get_metadata_pool());
+
+  map<string, bufferlist> _set;
+  set<string> _rm;
+
+  unsigned max_write_size = mdcache->max_dir_commit_size;
+  unsigned write_size = 0;
+
+  char *dftstr = dfts.c_str();
+
+  auto commit_one = [&](bool header=false) {
+    ObjectOperation op;
+
+    // don't create new dirfrag blindly
+    if (!_new)
+      op.stat(nullptr, nullptr, nullptr);
+
+    /*
+     * save the header at the last moment.. If we were to send it off before
+     * other updates, but die before sending them all, we'd think that the
+     * on-disk state was fully committed even though it wasn't! However, since
+     * the messages are strictly ordered between the MDS and the OSD, and
+     * since messages to a given PG are strictly ordered, if we simply send
+     * the message containing the header off last, we cannot get our header
+     * into an incorrect state.
+     */
+    if (header) {
+      bufferlist header;
+      encode(*fnode, header);
+      op.omap_set_header(header);
+    }
+
+    op.priority = op_prio;
+    if (!_set.empty())
+      op.omap_set(_set);
+    if (!_rm.empty())
+      op.omap_rm_keys(_rm);
+    mdcache->mds->objecter->mutate(oid, oloc, op, snapc,
+                                   ceph::real_clock::now(),
+                                   0, gather.new_sub());
+    write_size = 0;
+    _set.clear();
+    _rm.clear();
+  };
+
+  for (auto &key : stales) {
+    write_size += key.length();
+    _rm.emplace(key);
+
+    if (write_size >= max_write_size)
+      commit_one();
+  }
+
+  for (auto &k : to_remove) {
+    string key;
+    k.encode(key);
+    write_size += key.length();
+    _rm.emplace(std::move(key));
+
+    if (write_size >= max_write_size)
+      commit_one();
+  }
+
+  uint64_t off = 0;
+  bufferlist bl;
+  for (auto &item : to_set) {
+    string key;
+    item.key.encode(key);
+
+    encode(item.first, bl);
+    if (item.is_remote) {
+      bl.append('L');         // remote link
+      encode(item.ino, bl);
+      encode(item.d_type, bl);
+    } else {
+      bl.append('I');         // inode
+      encode(*item.inode, bl, item.features);
+      if (item.snaprealm)
+        encode(item.srnode, bl);
+      else
+        encode(bufferlist(), bl);
```

Comment:
This should be after `item.xattrs`.

Reply:
With this fixing, I am still getting the corrupt:

```
2020-09-20T22:33:26.922+0000 7f55734c1700 10 MDSIOContextBase::complete: 21C_IO_Dir_OMAP_Fetched
2020-09-20T22:33:26.922+0000 7f55734c1700 10 MDSContext::complete: 21C_IO_Dir_OMAP_Fetched
2020-09-20T22:33:26.922+0000 7f55734c1700 10 mds.0.cache.ino(0x10000000000) verify_diri_backtrace
2020-09-20T22:33:26.922+0000 7f55734c1700 10 mds.0.cache.dir(0x10000000000) _fetched header 274 bytes 1 keys for [dir 0x10000000000 /dir_1/ [2,head] auth v=0 cv=0/0 ap=1+0 state=1073741888|fetching f() n() hs=0+0,ss=0+0 | waiter=1 authpin=1 0x55b1381c0d00]
2020-09-20T22:33:26.922+0000 7f55734c1700 10 mds.0.cache.dir(0x10000000000) _fetched version 11
2020-09-20T22:33:26.922+0000 7f55734c1700 10  mds.0.cache.snaprealm(0x1 seq 1 0x55b1373a7900) have_past_parents_open [1,head]
2020-09-20T22:33:26.922+0000 7f55734c1700 10  mds.0.cache.snaprealm(0x1 seq 1 0x55b1373a7900) have_past_parents_open [1,head]
2020-09-20T22:33:26.922+0000 7f55734c1700 10  mds.0.cache.snaprealm(0x1 seq 1 0x55b1373a7900) have_past_parents_open [1,head]
2020-09-20T22:33:26.922+0000 7f55734c1700 10  mds.0.cache.snaprealm(0x1 seq 1 0x55b1373a7900) get_snaps  (seq 1 cached_seq 1)
2020-09-20T22:33:26.922+0000 7f55734c1700 10  mds.0.cache.snaprealm(0x1 seq 1 0x55b1373a7900) have_past_parents_open [1,head]
2020-09-20T22:33:26.922+0000 7f55734c1700 10 mds.0.cache.dir(0x10000000000)  snap_purged_thru 0 < 1, snap purge based on
2020-09-20T22:33:26.922+0000 7f55734c1700 10  mds.0.cache.snaprealm(0x1 seq 1 0x55b1373a7900) have_past_parents_open [1,head]
2020-09-20T22:33:26.922+0000 7f55734c1700 20 mds.0.cache.dir(0x10000000000) _fetched pos 0 marker 'I' dname 'file_a [2,head]
2020-09-20T22:33:26.922+0000 7f55734c1700 20 mds.0.cache.dir(0x10000000000) lookup (head, 'file_a')
2020-09-20T22:33:26.923+0000 7f55734c1700  0 log_channel(cluster) log [WRN] : Corrupt dentry 'file_a' in dir frag 0x10000000000: End of buffer(/dir_1)
2020-09-20T22:33:26.923+0000 7f55734c1700 10 mds.0.cache.dir(0x10000000000) go_bad_dentry file_a
2020-09-20T22:33:26.923+0000 7f55734c1700 10 mds.0.cache.dir(0x10000000000) pre_dirty 12
2020-09-20T22:33:26.923+0000 7f55734c1700 10 mds.0.cache.dir(0x10000000000) _mark_dirty (was clean) [dir 0x10000000000 /dir_1/ [2,head] auth v=12 cv=11/11 ap=1+0 state=1073741825|complete f(v0 m2020-09-20T22:33:14.784637+0000 1=1+0) n(v0 rc2020-09-20T22:33:15.275168+0000 b8388608 1=1+0) hs=0+0,ss=0+0 | waiter=1 authpin=1 0x55b1381c0d00] version 12
```

from https://pulpito.ceph.com/xiubli-2020-09-20_14:07:42-fs-wip-testing-20200920-1843-distro-basic-smithi/5452384

Reply:
With the "End of buffer" error, it should be the encoding issue, when decoding it it reached the buffer end, but it shouldn't in `_load_dentry()-->decode_bare()`.

Reply:
@ukernel 
BTW, do the following code could affect this issue ?

```
--- a/src/mds/CDir.cc
+++ b/src/mds/CDir.cc
@@ -2278,6 +2278,7 @@ void CDir::_omap_commit_ops(int r, int op_prio, version_t version, bool _new, bu
 
   uint64_t off = 0;
   bufferlist bl;
+  using ceph::encode;
   for (auto &item : to_set) {
     string key;
     item.key.encode(key);
@@ -2463,6 +2464,7 @@ void CDir::_parse_dentry(CDentry *dn, dentry_commit_item &item,
     item.inode = in->inode;
     if (in->inode->is_symlink())
       item.symlink = in->symlink;
+    using ceph::encode;
     encode(in->dirfragtree, bl);
     item.xattrs = in->xattrs;
     item.old_inodes = in->old_inodes;
```
The original code has the `using ...`

Reply:
Thanks @ukernel for pointing out the root cause.
It dues to the `snamerealm->srnode` encoding issue. It should be encoded to one `bufferlist` first and then encode the `bufferlist` to another `bufferlist`.